### PR TITLE
Add a testing facade API to streamline our keyhive testing strategy for higher-level tests

### PR DESCRIPTION
**Pull request by jtfmumm**
See `keyhive_core/tests/facade/README.md` for the details.

The code in this PR basically falls into 2 categories:

1. The facade API implementation in `keyhive_core/tests/facade/mod.rs` (roughly 1500 lines),
2. New test files where existing keyhive tests have been ported over to use the facade API.

To make reviewing this PR more manageable, I'd say look at the new `README.md`, review the facade implementation, and spot-check the actual tests. 

This PR does not (yet) include deletion of those old tests.

### CODE REVIEW

File: keyhive_core/tests/facade/mod.rs [hunk 1]
```diff
@@ -0,0 +1,1560 @@
+//! A testing facade for `Keyhive`.
```
Comment (expede) [inline keyhive_core/tests/facade/mod.rs:1]: Style: we should use the modern module layout (`facade.rs`). `mod.rs` is deprecated in the community style

**Comment by expede** [comment]
Retaining high level regression tests is certainly a nice-to-have, though as you note, there's a nontrivial chance that we alter several assumptions. I don't think that's a weakness of the project per se; this is R&D we should expect our assumptions to change. As such: I'm not yet convinced that this strategy will achieve the stated goals in the ADR. We may get the benefits with copy/paste over a facade that locks us into certain abstractions — not say that for sure, just flagging my scars from prior attempts at the similar!

AFAICT, this is largely test set up and wrapping existing APIs. I think there is real value to having manually audited the tests, and we should bake that into our process more.

I'm not sure that this style of tooling is what's going to help us have effective tests in the long term, especially as test needs won't fit neatly into the context as it's set up — my bias is having seen these kinds of things bit rot as test needs evolve and stack edge cases.

:sparkles: Importantly, :sparkles: regardless of the test strategy, I think the biggest win from this exercise is showing which APIs we're missing from the non-test code. We've talked about higher level APIs for a while but haven't actually done them as we've been busy with other things. With few exceptions, we should be able to express most (all except malicious scenarios?) of the scenario setup code in the main production code. At least to me this suggests that we could move these to become convenience APIs / the main way that people interact with the system.

Thoughts?

File: keyhive_core/tests/facade/mod.rs [hunk 2]
```diff
@@ -0,0 +1,1560 @@
...
+            + s.pending_delegated
+            + s.pending_revoked) as usize)
+    }
+
+    /// Sends every individual what every other has not yet sent them, repeating until a
+    /// round changes nothing.
+    ///
+    /// Returns an error if the state has not settled after eight rounds.
+    pub async fn sync_all_unsent(&mut self) -> Result<()> {
+        const MAX_ROUNDS: usize = 8;
```
Comment (FintanH) [inline keyhive_core/tests/facade/mod.rs:1300]: I have a feeling this constant will trip someone up some day. I would suggest a method that takes the number of rounds, e.g. `sync_all_unsent_within(&self, number_of_round: usize) -> Result<()>`, and this could delegate to that with `MAX_ROUNDS: usize = 8`, and a comment to say why 8 is a perfectly good number here.

File: keyhive_core/tests/facade/mod.rs [hunk 3]
```diff
@@ -0,0 +1,1560 @@
...
+type AgentHandle = Agent<future_form::Sendable, MemorySigner, [u8; 32], NoListener>;
+type MemberedHandle = Membered<future_form::Sendable, MemorySigner, [u8; 32], NoListener>;
+type DocHandle = Arc<Mutex<Document<future_form::Sendable, MemorySigner, [u8; 32], NoListener>>>;
+type ContentStore = MemoryCiphertextStore<[u8; 32], Vec<u8>>;
+
+pub type Result<T> = std::result::Result<T, TestError>;
+
+/// Why an operation was refused.
+#[derive(Debug, Error)]
+pub enum TestError {
```
Comment (FintanH) [inline keyhive_core/tests/facade/mod.rs:58]: I think it's important to note for maintainers that these are errors we are expecting to match on in the testing context. New variants may need to be added to be able to do that.

I would suggest adding a comment to make note of that.

**Review by FintanH — COMMENTED** [review]
Just a couple of things, since you mentioned (out-of-band) that you're changing it.

I will say that it did make reading the tests a lot easier! I read through some of the access control tests and they were quite clear – modulo my lack of knowledge for some Keyhive properties.

**Comment by jtfmumm** [comment]
Closed in favor of #222
